{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5b304ff8-3327-4e29-8392-5b22bce94cb0",
   "metadata": {
    "tags": []
   },
   "source": [
    "# PDF to Text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "07aed559-b9c9-461e-b406-776b9a735287",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pdfplumber in /opt/conda/lib/python3.9/site-packages (0.6.1)\n",
      "Requirement already satisfied: Pillow>=9.1 in /opt/conda/lib/python3.9/site-packages (from pdfplumber) (9.1.0)\n",
      "Requirement already satisfied: Wand>=0.6.7 in /opt/conda/lib/python3.9/site-packages (from pdfplumber) (0.6.7)\n",
      "Requirement already satisfied: pdfminer.six==20220319 in /opt/conda/lib/python3.9/site-packages (from pdfplumber) (20220319)\n",
      "Requirement already satisfied: cryptography in /opt/conda/lib/python3.9/site-packages (from pdfminer.six==20220319->pdfplumber) (36.0.2)\n",
      "Requirement already satisfied: chardet in /opt/conda/lib/python3.9/site-packages (from pdfminer.six==20220319->pdfplumber) (4.0.0)\n",
      "Requirement already satisfied: cffi>=1.12 in /opt/conda/lib/python3.9/site-packages (from cryptography->pdfminer.six==20220319->pdfplumber) (1.15.0)\n",
      "Requirement already satisfied: pycparser in /opt/conda/lib/python3.9/site-packages (from cffi>=1.12->cryptography->pdfminer.six==20220319->pdfplumber) (2.21)\n"
     ]
    }
   ],
   "source": [
    "!pip install pdfplumber"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1e206ec9-ab3c-465a-997b-ddb207f13b36",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pdfplumber"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "92e25fc9-8b9e-49f3-b574-369b95840505",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "PatternRecognitionLetters\n",
      "journalhomepage: www.elsevier.com\n",
      "Fast computation of mutual information in the frequency domain with applications to\n",
      "global multimodal image alignment\n",
      "JohanO¨fverstedta,∗∗,JoakimLindblada,NatasˇaSladojea\n",
      "1 aDepartmentofInformationTechnology,UppsalaUniversity,La¨gerhyddsva¨gen2,75237Uppsala,Sweden\n",
      "2\n",
      "0\n",
      "2 ABSTRACT\n",
      " \n",
      "n\n",
      "Multimodalimagealignmentistheprocessofﬁndingspatialcorrespondencesbetweenimagesformed\n",
      "u\n",
      "J bydiﬀerentimagingtechniquesorunderdiﬀerentconditions, tofacilitateheterogeneousdatafusion\n",
      "8  andcorrelativeanalysis.Theinformation-theoreticconceptofmutualinformation(MI)iswidelyused\n",
      "2 as a similarity measure to guide multimodal alignment processes, where most works have focused\n",
      "  on local maximization of MI that typically works well only for small displacements; this points to\n",
      " \n",
      "]\n",
      "a need for global maximization of MI, which has previously been computationally infeasible due to\n",
      "V\n",
      "the high run-time complexity of existing algorithms. We propose an eﬃcient algorithm for comput-\n",
      "C\n",
      "ingMIforalldiscretedisplacements(formalizedasthecross-mutualinformationfunction(CMIF)),\n",
      ".\n",
      "s whichisbasedoncross-correlationcomputedinthefrequencydomain. Weshowthatthealgorithm\n",
      "c\n",
      "isequivalenttoadirectmethodwhileasymptoticallysuperiorintermsofrun-time. Furthermore,we\n",
      "[\n",
      "  propose a method for multimodal image alignment for transformation models with few degrees of\n",
      " \n",
      "1 freedom (e.g. rigid) based on the proposed CMIF-algorithm. We evaluate the eﬃcacy of the pro-\n",
      "v\n",
      "posed method on three distinct benchmark datasets, of aerial images, cytological images, and histo-\n",
      "9\n",
      "logical images, and we observe excellent success-rates (in recovering known rigid transformations),\n",
      "9\n",
      "6 overall outperforming alternative methods, including local optimization of MI as well as several re-\n",
      "4 cent deep learning-based approaches. We also evaluate the run-times of a GPU implementation of\n",
      "1 the proposed algorithm and observe speed-ups from 100 to more than 10,000 times for realistic im-\n",
      ".\n",
      "6 age sizes compared to a GPU implementation of a direct method. Code is shared as open-source at\n",
      "0 github.com/MIDA-group/globalign.\n",
      "1 ©2021ElsevierLtd.Allrightsreserved.\n",
      "2\n",
      ":\n",
      "v\n",
      "i\n",
      "X\n",
      "1. Introduction containsmall/thinstructuresthatarediﬃculttomatchwithout\n",
      "r\n",
      "a relying on ﬁducial markers or time-consuming manual inter-\n",
      "Multimodal image alignment (also known as registration)\n",
      "vention[2].\n",
      "[1], involves ﬁnding correspondences between images formed\n",
      "by diﬀerent imaging techniques or under diﬀerent conditions. Monomodal(orunimodal)alignmentcanbeaddressedbya\n",
      "number of existing techniques, often grouped into (i) feature-\n",
      "The goal is often to enable data fusion of the heterogeneous\n",
      "based methods, such as SIFT [3], where the focus is on ﬁnd-\n",
      "information of the sources involved. Processes closely related\n",
      "ing correspondences between distinct feature points detected\n",
      "toregistrationarepatchretrievalandtemplatematching. Mul-\n",
      "in the images, and (ii) intensity-based methods, such as [4],\n",
      "timodal image alignment can be a very challenging problem,\n",
      "wherethealignmentisguidedbysimilaritybetweenthewhole\n",
      "duetogreatdissimilaritiesoftheinvolvedmodalities. Onearea\n",
      "image functions. Multimodal alignment is more complex and\n",
      "wherefullyautomatedapproacheshavebeenlackingisinalign-\n",
      "challenging,sincesimilarityinappearancecannotbeexpected\n",
      "mentofmicrographs,andinparticularcorrelativemicroscopy,\n",
      "for corresponding structures. This reduces the applicability\n",
      "where image modalities are often highly distinct and images\n",
      "of feature-based methods. Instead, intensity-based methods\n",
      "guidedbylocaloptimizationofsimilaritymeasureswhichrely\n",
      "∗∗Correspondingauthor: on some statistics, such as mutual information (MI) [5, 6] or\n",
      "e-mail:johan.ofverstedt@it.uu.se(JohanO¨fverstedt) modalityindependentneighbourhooddescriptors[7],arewell-\n",
      "\n",
      "2\n",
      "A A (Quantized)\n",
      "CMIFs Aligned\n",
      "Quantize Compute Max CMIF A and T(B)\n",
      "(k-means) CMIFs gives T\n",
      "B B (Quantized)\n",
      "Fig.1: Illustrationofthemainstepsoftheproposedimagealignmentmethod. TheinputconsistsoftwounalignedhistologicalimagesfromSHG(modalityA)\n",
      "andBFimaging(B).Theyarebothquantizedinto16categoricallabelsusingk-meansclustering(illustratedbypseudo-coloring)followedbycomputationofdense\n",
      "CMIFmaps(MIoveralldiscretedisplacements)foreachconsideredrotationangle. ThetransformationwiththemaxCMIFisidentiﬁed;itprovidesthedesired\n",
      "alignment.Attheendofthepipelineweobservethetransformedoverlappingimageswhicharesuccessfullyaligned.\n",
      "establishedtoolsforthistask. Anotherclassofmethodsrelyon run-time of the fast CMIF-algorithm in comparison to the di-\n",
      "reducingamultimodalalignmenttasktoamonomodalone,by recthistogram-basedalgorithm. Weobservespeed-upsranging\n",
      "(typicallylearning-based)transformationoftheimagemodali- from hundreds of times to more than 10,000 times for practi-\n",
      "tiestoacommonmodality[8,9],forwhichhigh-performance callyrelevantimagesizes.\n",
      "monomodalmethodsareapplicable. Ourﬁfthcontributionisanevaluationoftheproposedglobal\n",
      "IthasbeenobservedthatMIexhibitshighperformancewhen rigidalignmentmethodcomparedwithstate-of-the-artmethods\n",
      "usedasasimilaritymeasureinlocaloptimizationframeworks onthreedistinctandpubliclyavailablemultimodalrigidalign-\n",
      "if the displacements to recover are small, but struggles for mentbenchmarkdatasets,followingtheprotocolof[9].\n",
      "largerdisplacements[4,6,9]. Thishighlightstheneedforfast The two proposed algorithms are implemented in PyTorch\n",
      "globaloptimizationofMI.However,existingmethods[10,11] [13] for accelerated computation on general-purpose graphics\n",
      "exhibit run-time complexities which make them unsuited for processing units (GPGPUs), and the code is shared as open-\n",
      "manypracticallyrelevantimagesizes. sourceatgithub.com/MIDA-group/globalign.\n",
      "We here present two algorithms. The ﬁrst algorithm, and\n",
      "main contribution of this work, eﬃciently computes MI be- 2. MIinimageprocessing\n",
      "tween two images for all possible discrete displacements on a\n",
      "rectangular domain. The output corresponds to a generaliza- MI, introduced in [14], is an information-theoretic measure\n",
      "tion of the, in the ﬁeld of 1D (EEG) signal analysis, existing quantifying the mutual dependence between two random vari-\n",
      "notion of cross-mutual information function (CMIF) [12] and ables. Itiscommonlyusedinimageprocessingasasimilarity\n",
      "we therefore use the same name in this work. We propose measurebetweenimages[15,16].\n",
      "a novel approach for eﬃcient computation of CMIF by using Given two discrete images A and B with domains XA ⊂ Zn\n",
      "cross-correlation (CC) in the frequency domain, providing an andXB ⊂Zn,intersectingonXAB,whereXAB ≠∅,withranges\n",
      "outputwhichisequivalenttoexhaustivelycomputingMIover AandB,MIisdeﬁnedas\n",
      "allpossiblediscretedisplacements.\n",
      "pAB(a,b)\n",
      "Our second contribution is a method that combines the fast I(A,B)=∑∑ pAB(a,b)log XAB , (1)\n",
      "XAB 2 pA (a)pB (b)\n",
      "CMIFcomputationwithagridsearchoveradditionaltransfor- a∈Ab∈B XAB XAB\n",
      "mationparameters(e.g.rotation)tofacilitateglobalmultimodal\n",
      "where pAB(a,b)denotestherelativefrequencyofvaluesaand\n",
      "image alignment for transformations with reasonably few de- b occurrXinABg jointly in X , and pA (a) denotes the marginal\n",
      "grees of freedom (rigid, aﬃne, etc.). Figure 1 illustrates the AB XAB\n",
      "relativefrequencyofvalueaoccurringinimageA,withinX .\n",
      "main steps of global rigid alignment by this method applied AB\n",
      "Thefrequencies pAB(a,b), pA (a), pB (b)canbecomputed\n",
      "onhistologicalimagesacquiredbysecondharmonicgeneration XAB XAB XAB\n",
      "usingmarginalandjointhistograms.\n",
      "(SHG)microscopy,andbright-ﬁeld(BF)microscopy.\n",
      "Alternatively,MIcanbeformulatedintermsofthemarginal\n",
      "Our third contribution is a theoretical analysis of the pro-\n",
      "andthejointentropies,\n",
      "posed CMIF-algorithm. We show that it is equivalent to a di-\n",
      "recthistogram-basedalgorithmforcomputationofMI.Further- I(A,B)=H +H −H , (2)\n",
      "A B AB\n",
      "more, we derive the asymptotic computational complexity of\n",
      "theproposedCMIF-algorithm,andshowthatitissubstantially wherethemarginalentropyofimageA,onXABisgivenby\n",
      "morecomputationallyeﬃcientthantheexistingmethods.\n",
      "H =−∑pA (a)log pA (a), (3)\n",
      "Our fourth contribution is an empirical evaluation of the A XAB 2 XAB\n",
      "a∈A\n",
      "\n",
      "3\n",
      "analogousdeﬁnitionholdsforimageB,andthejointentropyis 3. Method\n",
      "givenby\n",
      "We propose a fast and exact algorithm for computing MI\n",
      "HAB=−∑∑ pXAABB(a,b)log2(pXAABB(a,b)). (4) forallpossiblediscretedisplacementsonarectangulardomain\n",
      "a∈Ab∈B (Eq. (5)), enabling fast global translation-based alignment us-\n",
      "CMIF describes the MI between two images subject to dis- ing MI maximization. We show the equivalence to a direct\n",
      "placementχ∈Zn(generalizationtonDofdeﬁnitionin[12]), histogram-based method and we analyze the asymptotic com-\n",
      "CMIF(χ;A,B)=I(A,T [B]), (5) putationalcomplexity. Furthermore,weproposeaglobalmul-\n",
      "χ\n",
      "timodalimagealignmentmethodbasedontheproposedCMIF-\n",
      "where Tχ[B] denotes the image B translated as Tχ[B](x) = algorithm.\n",
      "B(x+χ). CMIFcanalsobeformulatedintermsofentropies,\n",
      "similartoEq.(2),forentropiesdeﬁnedasfunctionsofdisplace- 3.1. Notationandbasicdeﬁnitions\n",
      "ments,HA(χ)=HA,HB(χ)=HTχ[B],andHAB(χ)=HATχ[B]. Consider pairs of images, A∶XA→A and B∶XB→B, where\n",
      "X ⊆X ⊂Zn, togetherwithcorrespondinggivenregionofin-\n",
      "2.1. ImagealignmentbyMImaximization B A\n",
      "terest masks, M ∶X →{0,1} and M ∶X →{0,1} indicating\n",
      "A A B B\n",
      "AnimagealignmentprocessbasedonMImaximizationcan\n",
      "the user-deﬁned part to be included in the computation of MI.\n",
      "̂\n",
      "beexpressedasﬁndingatransformationT suchthat\n",
      "Withoutlossofgenerality,weassumethatX andX arerect-\n",
      "A B\n",
      "T̂=argmaxI(A,T(B)), (6) angular subsets of Zn, since any other subset can be obtained\n",
      "T∈Ω through appropriately deﬁned masks MA and MB. Let XS de-\n",
      "where A is the reference image, and T(B) denotes the geo- notethesetofdisplacementsχforwhichtheentiredomainof\n",
      "metricallytransformedﬂoatingimage,wheretransformationT theshiftedimageBintersectswiththedomainofimageA.\n",
      "is applied to image B to warp it into the space of A, and Ω TheCC,forreal-valuedfunctions f andgdeﬁnedonZn, is\n",
      "is a chosen set of transformations. Solutions to (6) are com- givenby\n",
      "monlysoughtlocally,usinggradient-basedmethods[5,15–17], (f ⋆g)(χ)= ∑ f(x)g(x+χ). (7)\n",
      "orrecentlythroughaGibbs-samplingprocess[18],orglobally x∈Zn\n",
      "throughcomputationofMIforeachdiscretedisplacement[10]. Let La[A]denote an indicator function (level set) on XA, of\n",
      "Use of local methods requires a good initial guess since MI imageAbeingequaltoa∈AandwithinthemaskimageMA,\n",
      "typicallyexhibitsalargenumberoflocalmaxima[6]. MIcan\n",
      "1, for M (x)=1 and A(x)=a\n",
      "be combined with gradient information to improve the perfor- L [A](x)={ A }. (8)\n",
      "a 0, otherwise\n",
      "mancewhenthemodalitiesexhibitsimilarityofvariation[19].\n",
      "To reduce the dependence on the size of the images and their\n",
      "3.2. AlgorithmforcomputingCMIFinthefrequencydomain\n",
      "overlap,anormalizedMI(NMI)[20]wasdeveloped.\n",
      "Fast computation of MI is critical in many applications, in Contrarytothedirectmethod,where,givenaparticulardis-\n",
      "particular when MI is used as a similarity measure for align- placement χ, histograms are computed for all image values a\n",
      "menttasks,duetothelargedataandnumberofsimilaritycom- andb, weinsteadreordertherequiredoperations; givenasin-\n",
      "putations involved. Stochastic sub-sampling is commonly uti- gleimagevaluea,(or)b,orapairofvalues(a,b),thenumber\n",
      "lized together with gradient-based local optimization methods ofoccurrencesofthevalue/pair,andthecorrespondingentropy\n",
      "[5,6]. Otherworksincludespeedinguplocaloptimizationap- contribution,iscomputedforallχ∈XS.\n",
      "proachesbyusingGPU-basedhistogrammethods[11,21]. ComputationofthejointhistogramentriesCAB(χ;A,B)can\n",
      "a,b\n",
      "beexpressedas\n",
      "2.2. Globaloptimizationoverdiscretedisplacements\n",
      "Eﬃcient methods for global optimization over discrete dis- CaA,bB(χ;A,B)=(La[A]⋆Lb[B])(χ), (9)\n",
      "placements (based on fast convolutions performed in the fre-\n",
      "and the marginal histogram entries CA(χ;A,B) and\n",
      "quencydomainoronuseofintegralimages)existforabsolute a\n",
      "CB(χ;A,B)respectively,canbeexpressedas\n",
      "mean diﬀerences, mean square diﬀerences [22], CC and nor- b\n",
      "malizedcross-correlation[23],while(tothebestofourknowl- CA(χ;A,B)=(L [A]⋆M )(χ),\n",
      "a a B\n",
      "edge)sofarnotforMI[6,10,11,24]. (10)\n",
      "CB(χ;A,B)=(M ⋆L [B])(χ).\n",
      "Adirect,histogram-based,approachforcomputingCMIFfor b A b\n",
      "all displacements χ requires, for each χ (considered in isola-\n",
      "Complete histograms are computed by evaluating Eq. (9) and\n",
      "tion), toiterate over XATχ[B] and incrementthe joint histogram Eq.(10)forallimagevaluesa∈Aandb∈B. Anotherquantity\n",
      "bincorrespondingtothevaluesaandb(andanalogouslyforthe ofinterestfortheCMIFcomputationisamapN∶X →R rep-\n",
      "marginalhistograms).Finallytheentropies(Eq.(3)and(4))are S ≥0\n",
      "resentingthenumberofvalidpoints(pointswherebothmasks\n",
      "computed using relative histograms, which then directly give arenon-zero)forallχ∈X ,whichenablescomputationofrel-\n",
      "S\n",
      "theMIthroughEq.(2). Thismethodisdiscussedin[10]. Fur-\n",
      "ativehistograms. N isgivenby\n",
      "thermore, [11, 21] discuss maximization of MI by exhaustive\n",
      "search in the context of their fast approaches for GPU-based N(χ;A,B)=(M ⋆M )(χ)=\n",
      "A B\n",
      "histogramcomputation,andconcludethatsuchanapproachis ∑CA(χ,A,B)= ∑CB(χ,A,B). (11)\n",
      "toocostlytobeusefulinpractice. a b\n",
      "a∈A b∈B\n",
      "\n",
      "4\n",
      "The relative histograms obtained after normalizing Ca, Cb, 3.5. Complexityanalysis\n",
      "A B\n",
      "Ca,bby(pointwisedivisionby)Ncanbeinsertedasprobability\n",
      "AB Theworst-caserun-timecomplexityofthedirectapproachis\n",
      "functionsintotheshiftedversionsofEq.(3)andEq.(4),\n",
      "T (A,B)=O(∣X ∣∣X ∣+(∣A∣+∣B∣+∣A∣∣B∣)∣X ∣), (14)\n",
      "D B S S\n",
      "CA(χ;A,B) CA(χ;A,B)\n",
      "HA(χ)=−∑ Na(χ;A,B) log2 Na(χ;A,B) , given by its requirement of O(1) work for every element in\n",
      "a∈A image B for every χ ∈ X and O(1) work per histogram bin.\n",
      "CB(χ;A,B) CB(χ;A,B) S\n",
      "HB(χ)=−b∑∈B Nb(χ;A,B) log2 Nb(χ;A,B) , (12) Orenquthireeso,thinesrtheaadn,d,cocommppuutattaitoionnooffaCtCmoinstth1e+fr∣eAq∣u+en∣Bcy∣+do∣Am∣a∣Bin∣\n",
      "CAB(χ;A,B) CAB(χ;A,B) (forward and inverse) DFTs using the Fast Fourier Transform\n",
      "H (χ)=−∑∑ a,b log a,b ,\n",
      "AB N(χ;A,B) 2 N(χ;A,B) (FFT) algorithm [26] and element-wise (complex) multiplica-\n",
      "a∈Ab∈B tions,yieldinganasymptoticrun-timecomplexityof\n",
      "whichcanbeuseddirectlytocomputeCMIF(Eq.(5)). Round-\n",
      "T (A,B)=O((1+∣A∣+∣B∣+∣A∣∣B∣)∣X ∣log∣X ∣). (15)\n",
      "ingCA,CB,CABandNforallχtointegerswecircumventpos- F A A\n",
      "a b a,b\n",
      "sibleﬂoating-pointerrorsinthecomputationandreachanexact If ∣A∣ and ∣B∣ are treated as (small) constants, the proposed\n",
      "result. InEq.(12),wedeﬁnethat 00 =0and0log20=0. methodisasymptoticallyeﬃcientcf. thedirectapproach(w.r.t.\n",
      "thesizeoftheimages),exceptforcaseswithverysmall∣X ∣. It\n",
      "B\n",
      "hasbeenobserved[19]that32intensitybinsprovidedthebest\n",
      "3.3. GeneralizationtospatiallyweightedMI\n",
      "trade-oﬀbetween(i)ﬂexibilityanddetail,and(ii)insensitivity\n",
      "tonoise,forgradient-basedimagealignment.\n",
      "Spatially weighted MI (SWMI) [25] is a version of MI that\n",
      "associates each x ∈ X with a weight, modelling the relative ForverysmallimagesBorsetsofdisplacements∣XS∣,orfor\n",
      "A large value sets A and B, the direct method may be the bet-\n",
      "importanceofthecorrespondingimageelement(anditscontri-\n",
      "ter choice. In practice, small or medium sized value sets are\n",
      "bution to the relative frequencies). Our here proposed CMIF-\n",
      "usually acceptable and the image sizes are often such that the\n",
      "algorithm can be generalized to compute SWMI by replacing\n",
      "the binary mask M with a weight mask W ∶X → R , and proposedalgorithmisseveralordersofmagnitudesfaster.\n",
      "A A A ≥0\n",
      "modifyingthelevel-setformulationtoaweightedversion,\n",
      "3.6. Methodforglobalmultimodalalignment\n",
      "W (x), for A(x)=a We propose a method for image alignment by global MI\n",
      "LW[A](x)={ A }. (13)\n",
      "a 0, otherwise maximizationfortransformationswithfewdegreesoffreedom\n",
      "(e.g.rigidoraﬃne)bycombiningtheeﬃcientCMIF-algorithm\n",
      "proposedinSec.3.2,withgridsearchoverothertransformation\n",
      "Wemayfurthergeneralizethisapproachbyconsideringweight\n",
      "masks for both images A and B where the weight for a parameterssuchasrotationangle.Relyingonglobalsearch,the\n",
      "given x ∈ X and χ ∈ X is taken to be the product, methodcanﬁndtheglobalmaximumwithoutsmoothingofthe\n",
      "A S\n",
      "w(x)=W (x)⋅T [W ](x). images, which otherwise is commonly performed to increase\n",
      "A χ B\n",
      "thesizeoftheregionofattractionoftheglobalmaximum,but\n",
      "mayleadtoﬁndingasub-optimalsolution[6].\n",
      "3.4. Equivalenceofthecorrelation-basedalgorithmandthedi- Themethod,describedinAlg.1,performsglobalalignment\n",
      "rectmethod by taking a set of transformations Ω={T ,...,T } (e.g. se-\n",
      "1 k\n",
      "lected from a grid in the parameter space, or through random\n",
      "The proposed algorithm uses CC to compute CMIF ex- selection), warping the ﬂoating image (B) for each transfor-\n",
      "actly, and identically to the direct method; both involve com- mationusingnearest-neighbor(NN)interpolation,computinga\n",
      "puting contributions to the same discrete histogram counts CMIFmapusingtheproposedalgorithm,andlocatingthedis-\n",
      "N(χ,A,B), CA(χ,A,B), CB(χ,A,B) andCAB(χ,A,B) for all placement with highest MI. We use NN interpolation to com-\n",
      "a b a,b\n",
      "a,b and χ ∈ X . The CC (Eq. (7)) computes a sliding inner- pute T(B) since B is assumed to be categorical. A parameter\n",
      "S\n",
      "product between the two functions. For binary functions (im- γ∈[0,1]controlstherequiredfractionofoverlapcomparedto\n",
      "ages) f and g, Eq. (7) yields the number of elements x where the maximal observed overlap; γ is both used to select trans-\n",
      "f(x)=g(x+χ)=1, foragivenχ. Hence, theCCoflevel-set formations with suﬃcient overlap, and to control how large\n",
      "and level-set, level-set and mask, as well as mask and mask, padding is required. Finally, the resulting transformation is\n",
      "givethehistogramentries,foreachχ,asgivenbyEq.(9),(10) takenasthecompositionofthetransformationfromΩandthe\n",
      "and (11). These inner-products provide all the quantities re- T ,χ∈X whichleadstothehighestMIwithrequiredoverlap.\n",
      "χ S\n",
      "quiredforexactcomputationoftheshiftedentropies(Eq.(12)), TheﬁrststepofAlg.1istoquantizetheimagesintosuitably\n",
      "andthusforCMIF.Themaindistinctionbetweentheproposed smallnumberoflevelsusingsomeappropriatequantizationap-\n",
      "algorithm and the direct method is the order in which the his- proach(unlesstherangesarealreadysuitablediscreterepresen-\n",
      "togram counts are computed; the direct method computes the tations). Weusek-meansclustering,whichhasbeenshownto\n",
      "histogram entries (N(χ;A,B),CA(χ,A,B), CB(χ,A,B) for all yieldmoreeﬃcientutilizationofthediscretebinsforMI-based\n",
      "a b\n",
      "a,b,foraﬁxedχ,andtheproposedapproachcomputesthehis- imagealignment[17]thanequisizedbinning. Furthermore, k-\n",
      "togramentriesforallχ,forﬁxeda,b,beforeproceeding. means clustering enables the direct application of the method\n",
      "\n",
      "5\n",
      "Algorithm1ImagealignmentbyglobalMImaximization cytological image dataset consist of 864 and 5040 images re-\n",
      "Input: A∶XA→RmA,MA∶XA→{0,1}, spectively, and are both divided into three distinct folds; the\n",
      "B∶XB→RmB,MB∶XB→{0,1}, Ω∈{T1,...,Tk}, γ∈[0,1]. histological image dataset consists of a single set of 536 im-\n",
      "Output: I∈R≥0,T̂∶Rn→Rn. ages. Theaerialandcytologicalimagesareofsize300×300,\n",
      "1: A←QUANTIZE(A),B←QUANTIZE(B) andthehistologicalimagesareofsize834×834.\n",
      "2: A,MA←ZERO-PAD(A,MA;XB,1−γ)\n",
      "3: I←0,T̂←IDENTITY 5.1. Run-timeanalysis\n",
      "4: forT ∈Ωdo\n",
      "5: N ←γmaxN(χ;A,T(B)) We investigate the run-times of the proposed CMIF-\n",
      "χ∈XS algorithm,comparedtoadirecthistogram-basedalgorithm,im-\n",
      "6: χ← argmax CMIF(χ,A,T(B)) plemented in Python/PyTorch, using the built-in PyTorch im-\n",
      "χ∈XSs.t.N(χ;A,T(B))≥N\n",
      "7: IT ←CMIF(χ,A,T(B)) plementationforcomputingimagehistogramsontheGPU.The\n",
      "8: ifIT >Ithen algorithms implemented in PyTorch are very similar to meth-\n",
      "9: I←IT,T̂←T○Tχ ods 1 and 2 in [11], where the method is chosen dynamically\n",
      "10: endif basedonthenumberofrequestedbins,andthusenablingafair\n",
      "11: endfor comparisonoftheproposedmethodandthedirectmethod.The\n",
      "sort-and-countalgorithmforcomputingthehistograms[11]has\n",
      "beenshowntoberelevantmostlyforlargernumberofbins,and\n",
      "to multi-channel images, including image pairs with diﬀerent\n",
      "thereforewedonotconsiderithere,giventhatweobservethat\n",
      "numbersofchannels(denotedm andm ). Forthiswork, we\n",
      "A B smallnumberofbinsisshowntoworkwellinthisglobalopti-\n",
      "use the mini-batch k-means algorithm [27] which is fast and\n",
      "mizationcontext,asseeninSec.5.2.2.\n",
      "scalable. Wedenotekthenumberofclustersusedforquantiza-\n",
      "tionofbothimages,suchthatk=∣A∣=∣B∣.\n",
      "5.1.1. Experimentalsetup\n",
      "In Alg. 1, the reference image A and its mask M are zero-\n",
      "A\n",
      "paddedwith⌈∣X ∣(1−γ)⌉units(beforeandafter)along(each) We select one of the histological image pairs at ran-\n",
      "axisi,where∣XB∣idenotesthesize(diameter)ofthe(rectangu- dom, and crop/pad it to various sizes such that the\n",
      "Bi\n",
      "side-lengths of the reference image are a power of two\n",
      "lar)domainofimageBalongaxisi.\n",
      "{128,256,512,1024,2048,4096} and the other (ﬂoating) im-\n",
      "For the task of global rigid alignment we run the method\n",
      "age is of half the size along each dimension, which is the\n",
      "twice: (i)ﬁrstusinggridsearchtoreachacoarsealignmentat\n",
      "scenarioencounteredwhenaligningequallysizedimageswith\n",
      "anangleθ,(ii)followedbyan(optional)reﬁnementstep,where\n",
      "Alg.1andγ=0.5. WecomputetheCMIF-mapforallχ∈X .\n",
      "randomsearchisemployedinanintervalaroundthebestangle S\n",
      "foundbytheﬁrstgridsearch,samplinguniformlyfromthein- ExperimentsarerunonaNvidiaGeForceGTX2080GPU,\n",
      "terval[θ− 2π,θ+ 2π],where∣Ω∣denotesthenumberofangles with11GBofmemory.\n",
      "∣Ω∣ ∣Ω∣ Wealsoapplytheimagealignmentmethod(Alg.1)oncyto-\n",
      "intheﬁrstgridsearchstage.\n",
      "logicalandhistologicalimages,andmeasuretherun-timefora\n",
      "Bothimagealignmenttasks,wherethereferenceandﬂoating\n",
      "numberofconﬁgurations,including50,100,and200gridsteps\n",
      "images are similar in size, as well as patch retrieval (template\n",
      "(with 32 additional reﬁnement angles), with k ∈ {8,16,32},\n",
      "matching) tasks, where the reference image is larger than the\n",
      "withk-meansbatchsizeof1000andmaxiterations25.\n",
      "ﬂoatingimage,canbeeﬃcientlysolveddirectlybyAlg.1.\n",
      "5.1.2. Results\n",
      "4. Implementation\n",
      "The results of the run-time performance experiment regard-\n",
      "ing computation of a complete CMIF-map are summarized in\n",
      "We implement both the proposed CMIF-algorithm, as\n",
      "Fig.2. Weobservethattheproposedalgorithmissubstantially\n",
      "well as the global image alignment procedure (Alg. 1) in\n",
      "Python/PyTorch as a way to utilize parallel processing on a (between 100 times to more than 10000 times) faster than the\n",
      "directmethod,andthediﬀerenceisparticularlynoteworthyfor\n",
      "GPU.Using Python3.8.8and PyTorch1.8.1, the methodruns\n",
      "lownumbersofbins,k∈{2,4,8}.\n",
      "almostentirelyontheGPU,includingimagewarping,padding,\n",
      "Table 1 shows the run-time for the global rigid alignment\n",
      "level-sets,andCC(FFTsandcomplexmultiplication).\n",
      "methodforanumberofrealisticconﬁgurations.Forthesmaller\n",
      "Mini-batchk-meansclusteringreliesontheprovidedimple-\n",
      "cytologicalimages,with8binsinparticular,weobserveperfor-\n",
      "mentationinthesklearnpackage[28],version0.24.1.\n",
      "mancecompatiblewithreal-timeapplications. Onerigidimage\n",
      "alignment for the histological dataset using the direct method\n",
      "5. Performanceanalysis withk=16and200angles+32reﬁnementangles,runsinap-\n",
      "proximately 19 hours (compared to 32.9s with the fast CMIF-\n",
      "Weconsiderthreedatasets([29]): (i)aerialimages[30]with\n",
      "algorithm).\n",
      "infrared(IR)asonemodalityandcolorimagesastheother,(ii)\n",
      "cytological images [31] with quantitative phase images as one\n",
      "5.2. Rigidimagealignment\n",
      "modality and ﬂuorescence images as the other, (iii) histologi-\n",
      "cal images [32] with SHG images as one modality and BF as To evaluate the eﬃcacy of the proposed image alignment\n",
      "the other, (illustrated in Fig. 1). The aerial image dataset and method, compared to several existing methods, we follow the\n",
      "\n",
      "6\n",
      "10000.0 Direct (k=64) 1.0 Aerial\n",
      "Direct (k=32) 0.9 Cytological\n",
      "Run-time (s) 1101001000....0000 DDPPPrrriiooorreepppccooottsss  ((eeekkddd==   (((18kkk6)===)631426))) uccess-rate00000.....45678 Histological\n",
      "0.1 Proposed (k=8) S0.3\n",
      "0.01 Proposed (k=4) 0.2\n",
      "0.001 Proposed (k=2) 0.1\n",
      "1282 2562 5122 10242 20482 40962 0.0\n",
      "Reference image size (px) 2 4 8 16 32 64\n",
      "k\n",
      "Fig.2:Run-timeoftheproposedCMIF-algorithmforvariousreferenceimage\n",
      "Fig.3:Therelationshipbetweensuccess-rateandchoiceofkonthethreeeval-\n",
      "sizesandchoicesofk, incomparisontothedirecthistogram-basedmethod,\n",
      "uationdatasets.\n",
      "implementedinPyTorchandexecutedonaGPU.Weobservethattheproposed\n",
      "CMIF-algorithmisseveralordersofmagnitudefasterthanthedirectmethod.\n",
      "The empirical run-times of the proposed algorithm exhibits a strong depen-\n",
      "1.0 Aerial\n",
      "dencyonk,asexpectedfromEq.(15). Theempiricalrun-timesforthedirect 0.9 Aerial (w. refinement)\n",
      "mTbeaetbtwlheeoe1dn:aRsrueucnrhe-tlciamhtioveiecilenysssfetocarobknled∈sw{f.8or.,rt.1g6klo,;3bw2a,le6roi4gb}isd.erivmeadgaetamligonstmaendtiﬀfoerreancneumofbe3r0o%f Success-rate000000......345678 CCHHyyiissttttoooollllooooggggiiiiccccaaaallll  ((ww..  rreeffiinneemmeenntt))\n",
      "0.2\n",
      "conﬁgurationsusingimagestakenfromthecytologicalandhistologicaldataset. 0.1\n",
      "0.0\n",
      "25 50 100 150 200 250 300\n",
      "Cytological k Histological k Angles\n",
      "8 16 32 8 16 32\n",
      "Fig.4: Therelationshipbetweensuccess-rateandthenumberofanglesinthe\n",
      "s 50 0.8 2.2 7.3 s 50 4.0 11.8 42.2 gridsearch,usingk=32,bothwith,orwithout,thesubsequentreﬁnementstep.\n",
      "e e\n",
      "gl 100 1.2 3.4 11.4 gl 100 6.1 18.9 67.9 Inbothcases,theperformanceincreasessubstantiallyupto150anglesforthe\n",
      "n n\n",
      "A 200 2.2 6.1 22.1 A 200 10.3 32.9 119 threedatasets.\n",
      "evaluation protocol of [9], comprising a benchmark for evalu- erations. Weevaluatethemethodonthethreedatasetsusinga\n",
      "atingtheeﬀectivenessof2Drigidimagealignmentmethodson set of angle counts {25,50,100,150,200,250,300}, with and\n",
      "three distinct datasets. The study includes comparison of four withoutreﬁnementwith32randomlyselectedangles,andmea-\n",
      "image-to-image (I2I) translation methods, as well as a state- surethesuccess-rateforeachconﬁguration.\n",
      "of-the-art contrastive learning method for modality transfer, We use circular masks, to avoid bias from presence or ab-\n",
      "whichareallusedtoattempttotransformthemultimodalalign- senceofthesignalcontentsinthecornersoftheimages.\n",
      "ment task into an easier monomodal alignment task, and are\n",
      "then combined with either a high-performance intensity-based\n",
      "5.2.2. Results\n",
      "method[4]orawell-knownfeature-basedmethod[3]. Finally,\n",
      "First we present the performance on the three evaluation\n",
      "localoptimization(gradient-based)MIisalsoincluded;itisap-\n",
      "datasets(asmeasuredbysuccess-rates)fordiﬀerentchoicesof:\n",
      "pliedtomultimodaldatadirectly. Eventhough2Drigidtrans-\n",
      "kinFig.3,andanglecountinFig.4;Thesuccess-rateincreases\n",
      "formation is among the simplest models, solving these align-\n",
      "substantiallyuptok=16,and150angles.\n",
      "ment tasks still poses a challenge in the presence of multiple\n",
      "Furthermore, we present the performance of the considered\n",
      "distinctmodalities. Itisalsoarealistictaskinmicroscopyand\n",
      "methods in comparison to a number of alternative methods in\n",
      "aerialsettingswhererelativescalebetweentheimagesmaybe\n",
      "Tab. 2. The four I2I translation-based methods are combined\n",
      "known a priori, or readily estimated, while arbitrary rotations\n",
      "intoasinglerow,wherethemaximumsuccess-ratealongeach\n",
      "canbeencountered.\n",
      "column is shown. The proposed method is the best choice on\n",
      "twoofthedatasets(thecytologicalandthehistological)where\n",
      "5.2.1. Experimentalsetup\n",
      "it exhibits a substantial improvement over existing methods.\n",
      "Following [9], we apply the proposed alignment method on\n",
      "CoMIRcombinedwithSIFTreaches100%success-rateonthe\n",
      "allthreedatasets. Theperformancemeasureisbasedontheav-\n",
      "Zurich dataset, while the proposed method delivers a success-\n",
      "erageEuclideandistancebetweenimagecornersconsideredas\n",
      "rateof99.8%,whichisbetterthanallincludedintensity-based\n",
      "landmarksinthereferenceimagespace,andthecorresponding\n",
      "methods.\n",
      "recovered landmarks of the aligned ﬂoating image. An align-\n",
      "mentisconsideredsuccessfuliftheerrorislessthan2%ofthe\n",
      "widthoftheimages[9]. Table2:Successrates(presentedaspercentages)oftheevaluatedmethodson\n",
      "rigidalignmenttasksonthreedatasets. Resultsontheaerialandcytological\n",
      "Animportantparameteroftheproposedmethodisk,which\n",
      "datasetsarepresentedasempiricalmean±std-devoverthethreefolds. Bold\n",
      "hasalargeimpactontherun-time(Eq.15)whilealsodirectly\n",
      "marksthebestresultoneachdataset.Usedparametersfortheproposedmethod\n",
      "inﬂuencing how detailed structures can be represented in the arek=32,γ=0.5,200equispacedanglesin[−π,π]plus32randomreﬁnement\n",
      "quantizedrepresentations. Werunthealignmenttaskforallthe angles.Fordetailsaboutthereferencemethods,andtheirconﬁguration,see[9].\n",
      "images in the included datasets (aerial, cytological, and histo- Dataset AerialData CytologicalData HistologicalData\n",
      "logical), with k ∈ {2,4,8,16,32,64}, with k-means batch size Method α-AMD SIFT α-AMD SIFT α-AMD SIFT\n",
      "I2I 80.2±3.9 98.3±0.5 71.1±5.8 24.4±6.2 28.4 0\n",
      "of100andmaxiterations100,andmeasurethesuccess-rates. CoMIR 91.8±7.7 100.0±0.0 68.0±14.0 72.5±7.1 81.3 59.3\n",
      "The number of angles for the grid search, as well as if in- LocalMI 68.9±3.8 89.9±3.0 47.8\n",
      "cluding the reﬁnement step or not, are also important consid- Proposed 99.8±0.4 99.4±0.4 91.2\n",
      "\n",
      "7\n",
      "6. DiscussionandConclusion [6] J.P.Pluim, J.A.Maintz, M.A.Viergever, Mutual-information-based\n",
      "registrationofmedicalimages: asurvey, IEEETrans.Med.Imag.22\n",
      "WepresentanovelfastCMIF-algorithmwhichcomputesMI (2003)986–1004.\n",
      "[7] M.P.Heinrich,M.Jenkinson,M.Bhushan,T.Matin,F.V.Gleeson,S.M.\n",
      "foralldiscretedisplacementsoftwoimages(similarlyascross-\n",
      "Brady,J.A.Schnabel, MIND:Modalityindependentneighbourhoodde-\n",
      "correlation). The algorithm works in nD. The proposed algo- scriptorformulti-modaldeformableregistration, Med.ImageAnal.16\n",
      "rithmenablesfastglobalalignment(fortransformationmodels (2012)1423–1435.\n",
      "withfewdegreesoffreedom),hasveryfewparameters,andis [8] N. Pielawski, E. Wetzer, J. O¨fverstedt, J. Lu, C. Wa¨hlby, J. Lindblad,\n",
      "N. Sladoje, CoMIR: Contrastive multimodal image representation for\n",
      "straightforwardtoconﬁgurefornewapplications. Quantization\n",
      "registration,in:Neur.Inf.Proc.Sys.,volume33,2020,pp.18433–18444.\n",
      "ofimagerangeswithk-meansclusteringprovideseﬃcientuse [9] J.Lu,J.O¨fverstedt,J.Lindblad,N.Sladoje, Isimage-to-imagetransla-\n",
      "ofalimitednumberofquantizationlevels,andenablesaligning tionthepanaceaformultimodalimageregistration?Acomparativestudy,\n",
      "imageswithdiﬀerentnumbersofchannels. arXivpreprintarXiv:2103.16262(2021).\n",
      "[10] F.Barrera,F.Lumbreras,A.D.Sappa, Multimodaltemplatematching\n",
      "We compare the run-time of the proposed CMIF-algorithm\n",
      "basedongradientandmutualinformationusingscale-space, in: IEEE\n",
      "with a direct histogram-method (using a fast GPU-based his- Int.Conf.ImageProcess.,IEEE,2010,pp.2749–2752.\n",
      "togram algorithm), and observe speed-ups ranging from hun- [11] R.Shams,P.Sadeghi,R.Kennedy,R.Hartley, Parallelcomputationof\n",
      "mutualinformationontheGPUwithapplicationtoreal-timeregistration\n",
      "dreds of times to more than 10,000 times for practically rele-\n",
      "of3dmedicalimages,Comp.Meth.Prog.Biomed.99(2010)133–146.\n",
      "vant image sizes, indicating the potential for real-time global\n",
      "[12] B.Pompe,P.Blidh,D.Hoyer,M.Eiselt, Usingmutualinformationto\n",
      "multimodalalignment. Thepresentedmethodsarehighlypar- measurecouplinginthecardiorespiratorysystem, IEEEEng.Med.Biol.\n",
      "allelizable; the work can readily be distributed over multiple Mag.17(1998)32–39.\n",
      "[13] A. Paszke, S. Gross, et al., Pytorch: An imperative style, high-\n",
      "GPUsorcomputers.\n",
      "performancedeeplearninglibrary, AdvancesinNeuralInformationPro-\n",
      "Furthermore, we evaluate the performance of the proposed cessingSystems32(2019)8026–8037.\n",
      "alignment method on three datasets, and compare it with both [14] C.E.Shannon,Amathematicaltheoryofcommunication,BellSys.Tech.\n",
      "gradient-based MI method, and recent methods which rely on Jour.27(1948)379–423.\n",
      "[15] W.M.WellsIII,P.Viola,H.Atsumi,S.Nakajima,R.Kikinis, Multi-\n",
      "deep learning. We observe excellent performance of the pro-\n",
      "modalvolumeregistrationbymaximizationofmutualinformation,Med.\n",
      "posed method on all three datasets, and we conclude that it is ImageAnal.1(1996)35–51.\n",
      "the overall top performing method, even outperforming state [16] F.Maes,A.Collignon,D.Vandermeulen,G.Marchal,P.Suetens, Mul-\n",
      "of the art deep learning-based methods. Finally, the proposed timodality image registration by maximization of mutual information,\n",
      "IEEETrans.Med.Imag.16(1997)187–198.\n",
      "methoddoesnotrequirealignedimagepairs(oranytraining),\n",
      "[17] Z.F.Knops,J.A.Maintz,M.A.Viergever,J.P.Pluim, Normalizedmu-\n",
      "andhavefewparameterstotune,whichareadvantagesincom- tualinformationbasedregistrationusingk-meansclusteringandshading\n",
      "parisontothedeeplearning-basedmethods. correction,Med.ImageAnal.10(2006)432–439.\n",
      "[18] M.Agn,K.VanLeemput,Fastnonparametricmutual-information-based\n",
      "Onelimitationoftheproposedmethodisthatitdoesnotsup-\n",
      "registrationanduncertaintyestimation, in:Unc.SafeUtil.Mach.Learn.\n",
      "port cost-eﬀective sub-pixel alignment. Future work could in-\n",
      "Med.Imag.Clin.Image-BasedProc.,Springer,2019,pp.42–51.\n",
      "volve application of a suitable reﬁnement step based on local [19] J.Pluim,J.Maintz,M.Viergever, Imageregistrationbymaximization\n",
      "optimizationtowardsthisend. ofcombinedmutualinformationandgradientinformation, IEEETrans.\n",
      "Med.Imag.19(2000)809–814.\n",
      "[20] C.Studholme, D.L.Hill, D.J.Hawkes, Anoverlapinvariantentropy\n",
      "measureof3dmedicalimagealignment,Patt.Rec.32(1999)71–86.\n",
      "Declarationofcompetinginterests\n",
      "[21] R.Shams,N.Barnes,Speedingupmutualinformationcomputationusing\n",
      "nvidiacudahardware,in:DICTA2007,IEEE,2007,pp.555–560.\n",
      "Theauthorsdeclarenocompetinginterests. [22] M.J.Atallah,Fasterimagetemplatematchinginthesumoftheabsolute\n",
      "valueofdiﬀerencesmeasure,IEEETrans.Im.Proc.10(2001)659–663.\n",
      "[23] D.-M.Tsai,C.-T.Lin,Fastnormalizedcrosscorrelationfordefectdetec-\n",
      "Acknowledgments tion,Patt.Rec.Lett.24(2003)2625–2631.\n",
      "[24] S.Paul,U.C.Pati, Acomprehensivereviewonremotesensingimage\n",
      "registration,Int.Jour.ofRemoteSensing42(2021)5400–5436.\n",
      "ThisworkwassupportedbytheWallenbergAI,Autonomous\n",
      "[25] S.B.Park,F.C.Rhee,J.I.Monroe,J.W.Sohn,Spatiallyweightedmutual\n",
      "Systems and Software Program (WASP) AI-Math initiative; informationimageregistrationforimageguidedradiationtherapy, Med.\n",
      "VINNOVA (MedTech4Health project 2017-02447); and the Phys.37(2010)4590–4601.\n",
      "[26] J.W.Cooley,J.W.Tukey, Analgorithmforthemachinecalculationof\n",
      "SwedishResearchCouncil(project2017-04385).\n",
      "complexfourierseries,Math.Comput.19(1965)297–301.\n",
      "[27] D.Sculley, Web-scalek-meansclustering, in: Proc.Int.Conf.World\n",
      "wideWeb,2010,pp.1177–1178.\n",
      "References\n",
      "[28] F.Pedregosa,Varoquaux,etal.,Scikit-learn:MachinelearninginPython,\n",
      "Jour.ofMach.Learn.Res.12(2011)2825–2830.\n",
      "[1] B.Zitova,J.Flusser,Imageregistrationmethods:asurvey,ImageVision [29] [dataset]J.Lu,J.O¨fverstedt,J.Lindblad,N.Sladoje,2021,Datasetsfor\n",
      "Comput.21(2003)977–1000. EvaluationofMultimodalImageRegistration, doi:10.5281/zenodo.\n",
      "[2] P.Paul-Gilloteaux,X.Heiligenstein,et.al,ec-clem:ﬂexiblemultidimen- 4587903.\n",
      "sionalregistrationsoftwareforcorrelativemicroscopies, Nat.Meth.14 [30] M.Volpi,V.Ferrari, Semanticsegmentationofurbanscenesbylearning\n",
      "(2017)102–103. localclassinteractions, in: 2015IEEEConf.Comp.Vis.andPatt.Rec.\n",
      "[3] D.Lowe,Objectrecognitionfromlocalscale-invariantfeatures,in:Proc. Works.,2015,pp.1–9.\n",
      "IEEEInt.Conf.Comp.Vis.,volume2,1999,pp.1150–1157vol.2. [31] T.Vicar,M.Raudenska,J.Gumulec,J.Balvan, TheQuantitative-Phase\n",
      "[4] J.O¨fverstedt,J.Lindblad,N.Sladoje,FastandRobustSymmetricImage DynamicsofApoptosisandLyticCellDeath,Sci.Rep.10(2020)1–12.\n",
      "RegistrationBasedonDistancesCombiningIntensityandSpatialInfor- [32] A.Keikhosravi,B.Li,Y.Liu,K.W.Eliceiri, Intensity-basedregistration\n",
      "mation,IEEETrans.ImageProcess.28(2019)3584–3597. ofbright-ﬁeldandsecond-harmonicgenerationimagesofhistopathology\n",
      "[5] P.Viola,W.M.WellsIII, Alignmentbymaximizationofmutualinfor- tissuesections,Biomed.Opt.Express11(2020)160.\n",
      "mation,Int.J.Comp.Vis.24(1997)137–154.\n"
     ]
    }
   ],
   "source": [
    "file = \"sample_pdfs/MutualInformation.pdf\"\n",
    "pdf = pdfplumber.open(file)\n",
    "parsed_text = \"\\n\\n\".join([page.extract_text() for page in pdf.pages])\n",
    "print(parsed_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac701e27-b1ef-41bd-942c-9a02a461b310",
   "metadata": {},
   "source": [
    "# Extract all URLs in a website"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fbf540ac-720b-4dd8-8dbb-1b8c4940f3e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "from time import time\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6139623b-44e7-405c-bf1d-10c1ca5bac7f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "class URLExplorer:\n",
    "\n",
    "    def __init__(self, base_url, is_interesting_url):\n",
    "        self.reference_dag = {}\n",
    "        self.base_url = base_url\n",
    "        self.unexplored_urls = [base_url]\n",
    "        self.is_interesting_url = is_interesting_url\n",
    "        self.timings = {}\n",
    "        self.failed_urls = []\n",
    "\n",
    "    def get_referenced_urls(self, url):\n",
    "        request = requests.get(url)\n",
    "        soup = BeautifulSoup(request.text, \"html.parser\")\n",
    "        return [\n",
    "            link.get(\"href\")\n",
    "            for link in soup.find_all(\"a\")\n",
    "            if link.get(\"href\") and self.is_interesting_url(link.get(\"href\"))\n",
    "        ]\n",
    "\n",
    "    def explore_url(self, url):\n",
    "        if url in self.failed_urls:\n",
    "            return []\n",
    "\n",
    "        try:\n",
    "            start = time()\n",
    "            referenced_urls = self.get_referenced_urls(url)\n",
    "            end = time()\n",
    "            self.timings[url] = end - start\n",
    "\n",
    "        except:\n",
    "            self.failed_urls.append(url)\n",
    "            return []\n",
    "\n",
    "        self.reference_dag[url] = referenced_urls\n",
    "        self.unexplored_urls += [\n",
    "            url for url in referenced_urls if url not in self.reference_dag.keys()\n",
    "        ]\n",
    "        if url in self.unexplored_urls:\n",
    "            self.unexplored_urls.remove(url)\n",
    "\n",
    "    def explore_full_website(self, limit=None):\n",
    "        with tqdm() as pbar:\n",
    "            while self.unexplored_urls:\n",
    "                n_failed = len(self.failed_urls)\n",
    "                n_explored = len(self.reference_dag.keys()) + n_failed\n",
    "                n_unexplored = len(self.unexplored_urls)\n",
    "                n_total = n_explored + n_unexplored\n",
    "                text = f\"We have currently explored {n_explored}/{n_total} URLs ({n_failed} failures\"\n",
    "                if n_failed > 0:\n",
    "                    text += f\",namely {self.failed_urls[-1]})\"\n",
    "                else:\n",
    "                    text += \")\"\n",
    "                pbar.set_description(text)\n",
    "                pbar.update(1)\n",
    "                self.explore_url(self.unexplored_urls.pop())\n",
    "                if len(self.reference_dag.keys()) > limit:\n",
    "                    return\n",
    "\n",
    "    def number_of_connections(self):\n",
    "        return sum([len(v) for v in self.reference_dag.values()])\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f\"{self.base_url} Explorer. {len(self.reference_dag.keys())} nodes explored with {self.number_of_connections()} connections\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a55a138d-7ebf-4b08-8234-d839d71d0aa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "https://arxiv.org/ Explorer. 0 nodes explored with 0 connections"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def is_interesting_url(url):\n",
    "    return all(\n",
    "        [\n",
    "            url.startswith(\"https://arxiv.org\"),\n",
    "            \"ct?url=\" not in url,\n",
    "            not url.startswith(\"https://arxiv.org/search\"),\n",
    "            not url.startswith(\"https://arxiv.org/format\"),\n",
    "        ]\n",
    "    )\n",
    "\n",
    "\n",
    "explorer = URLExplorer(\n",
    "    base_url=\"https://arxiv.org/\", is_interesting_url=is_interesting_url\n",
    ")\n",
    "explorer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d04e0afe-12f9-4493-b3c3-c50705d84935",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "We have currently explored 30/445 URLs (0 failures): : 49it [00:59,  1.21s/it]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "https://arxiv.org/ Explorer. 31 nodes explored with 731 connections"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "explorer.explore_full_website(limit=30)\n",
    "explorer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "56a28c07-e60b-4a39-afb4-c98eef273071",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('https://arxiv.org/pdf/2203.03556', 30.036967039108276),\n",
       " ('https://arxiv.org/covid19search', 1.6784424781799316),\n",
       " ('https://arxiv.org/', 0.9685404300689697),\n",
       " ('https://arxiv.org/support/moderation_help', 0.9072434902191162),\n",
       " ('https://arxiv.org/corr/subjectclasses', 0.7228176593780518),\n",
       " ('https://arxiv.org/about', 0.7123024463653564),\n",
       " ('https://arxiv.org/corr', 0.5931897163391113),\n",
       " ('https://arxiv.org/help/submit', 0.5829927921295166),\n",
       " ('https://arxiv.org/help/endorsement#request', 0.5820391178131104),\n",
       " ('https://arxiv.org/help', 0.580359697341919),\n",
       " ('https://arxiv.org/login', 0.5691497325897217),\n",
       " ('https://arxiv.org/help/contact', 0.568894624710083),\n",
       " ('https://arxiv.org/about/people/developers', 0.5512149333953857),\n",
       " ('https://arxiv.org/help/policies/privacy_policy', 0.5424740314483643),\n",
       " ('https://arxiv.org/help/web_accessibility', 0.5417683124542236),\n",
       " ('https://arxiv.org/help/license', 0.5416324138641357),\n",
       " ('https://arxiv.org/help/view', 0.5414862632751465),\n",
       " ('https://arxiv.org/help/email-protection', 0.5378584861755371),\n",
       " ('https://arxiv.org/licenses/nonexclusive-distrib/1.0/', 0.5334033966064453),\n",
       " ('https://arxiv.org/help/subscribe', 0.531207799911499),\n",
       " ('https://arxiv.org/help/registerhelp', 0.5296356678009033),\n",
       " ('https://arxiv.org/help/mathjax', 0.5281879901885986),\n",
       " ('https://arxiv.org/help/moderation', 0.52620530128479),\n",
       " ('https://arxiv.org/help/policies/instructions_for_submission',\n",
       "  0.5159857273101807),\n",
       " ('https://arxiv.org/help/policies/submission_agreement', 0.5152199268341064),\n",
       " ('https://arxiv.org/abs/2203.03556', 0.49268031120300293),\n",
       " ('https://arxiv.org/cookies', 0.4484565258026123),\n",
       " ('https://arxiv.org/user/register?submit=Register+for+the+first+time',\n",
       "  0.4467332363128662),\n",
       " ('https://arxiv.org/archive/cs', 0.4406931400299072),\n",
       " ('https://arxiv.org/licenses/nonexclusive-distrib/1.0/license.html',\n",
       "  0.41458749771118164),\n",
       " ('https://arxiv.org/user/lost_password', 0.41370558738708496)]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(explorer.timings.items(), key=lambda x: x[1], reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c5b48a3c-cb8c-4781-8e34-7b7e6973aeef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://arxiv.org/\n",
      "https://arxiv.org/help/web_accessibility\n",
      "https://arxiv.org/help/policies/privacy_policy\n",
      "https://arxiv.org/help/license\n",
      "https://arxiv.org/help/subscribe\n",
      "https://arxiv.org/help/contact\n",
      "https://arxiv.org/help\n",
      "https://arxiv.org/about\n",
      "https://arxiv.org/login\n",
      "https://arxiv.org/user/register?submit=Register+for+the+first+time\n",
      "https://arxiv.org/user/lost_password\n",
      "https://arxiv.org/support/moderation_help\n",
      "https://arxiv.org/licenses/nonexclusive-distrib/1.0/\n",
      "https://arxiv.org/cookies\n",
      "https://arxiv.org/help/view\n",
      "https://arxiv.org/help/email-protection\n",
      "https://arxiv.org/help/endorsement#request\n",
      "https://arxiv.org/help/moderation\n",
      "https://arxiv.org/help/policies/instructions_for_submission\n",
      "https://arxiv.org/help/policies/submission_agreement\n",
      "https://arxiv.org/licenses/nonexclusive-distrib/1.0/license.html\n",
      "https://arxiv.org/help/registerhelp\n",
      "https://arxiv.org/help/submit\n",
      "https://arxiv.org/corr/subjectclasses\n",
      "https://arxiv.org/corr\n",
      "https://arxiv.org/archive/cs\n",
      "https://arxiv.org/covid19search\n",
      "https://arxiv.org/abs/2203.03556\n",
      "https://arxiv.org/help/mathjax\n",
      "https://arxiv.org/about/people/developers\n",
      "https://arxiv.org/abs/2203.03556v1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None,\n",
       " None]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[print(x) for x in explorer.reference_dag.keys()]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
